// @material-ui/lab/legacy/DateRangePicker/makeDateRangePicker.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import * as React from 'react';
import { unstable_useThemeProps as useThemeProps } from '@material-ui/core/styles';
import { useUtils } from '../internal/pickers/hooks/useUtils';
import { useParsedDate } from '../internal/pickers/hooks/date-helpers-hooks';
import { defaultMinDate, defaultMaxDate } from '../internal/pickers/constants/prop-types';
import { makeValidationHook } from '../internal/pickers/hooks/useValidation';
import { usePickerState } from '../internal/pickers/hooks/usePickerState';
import { DateRangePickerView } from './DateRangePickerView';
import DateRangePickerInput from './DateRangePickerInput';
import { parseRangeInputValue, validateDateRange } from '../internal/pickers/date-utils';
import { jsx as _jsx } from "react/jsx-runtime";
var useDateRangeValidation = makeValidationHook(validateDateRange, {
  isSameError: function isSameError(a, b) {
    return b !== null && a[1] === b[1] && a[0] === b[0];
  }
});
export default function makeDateRangePicker(name, Wrapper) {
  var KeyboardDateInputComponent = DateRangePickerInput;
  var PureDateInputComponent = DateRangePickerInput;

  function WrapperComponent(props) {
    var disableCloseOnSelect = props.disableCloseOnSelect,
        cancelText = props.cancelText,
        clearable = props.clearable,
        clearText = props.clearText,
        DateInputProps = props.DateInputProps,
        DialogProps = props.DialogProps,
        displayStaticWrapperAs = props.displayStaticWrapperAs,
        inputFormat = props.inputFormat,
        okText = props.okText,
        onAccept = props.onAccept,
        onChange = props.onChange,
        onClose = props.onClose,
        onOpen = props.onOpen,
        open = props.open,
        PopperProps = props.PopperProps,
        todayText = props.todayText,
        value = props.value,
        wrapperProps = props.wrapperProps,
        other = _objectWithoutProperties(props, ["disableCloseOnSelect", "cancelText", "clearable", "clearText", "DateInputProps", "DialogProps", "displayStaticWrapperAs", "inputFormat", "okText", "onAccept", "onChange", "onClose", "onOpen", "open", "PopperProps", "todayText", "value", "wrapperProps"]);

    var TypedWrapper = Wrapper;
    return /*#__PURE__*/_jsx(TypedWrapper, _extends({
      clearable: clearable,
      clearText: clearText,
      DialogProps: DialogProps,
      PopperProps: PopperProps,
      okText: okText,
      todayText: todayText,
      cancelText: cancelText,
      DateInputProps: DateInputProps,
      KeyboardDateInputComponent: KeyboardDateInputComponent,
      PureDateInputComponent: PureDateInputComponent,
      displayStaticWrapperAs: displayStaticWrapperAs
    }, wrapperProps, other));
  }

  var rangePickerValueManager = {
    emptyValue: [null, null],
    parseInput: parseRangeInputValue,
    areValuesEqual: function areValuesEqual(utils, a, b) {
      return utils.isEqual(a[0], b[0]) && utils.isEqual(a[1], b[1]);
    }
  };
  var RangePickerWithStateAndWrapper = /*#__PURE__*/React.forwardRef(function RangePickerWithStateAndWrapper(inProps, ref) {
    var props = useThemeProps({
      props: inProps,
      name: name
    });

    var calendars = props.calendars,
        value = props.value,
        onChange = props.onChange,
        _props$mask = props.mask,
        mask = _props$mask === void 0 ? '__/__/____' : _props$mask,
        _props$startText = props.startText,
        startText = _props$startText === void 0 ? 'Start' : _props$startText,
        _props$endText = props.endText,
        endText = _props$endText === void 0 ? 'End' : _props$endText,
        passedInputFormat = props.inputFormat,
        _props$minDate = props.minDate,
        minDateProp = _props$minDate === void 0 ? defaultMinDate : _props$minDate,
        _props$maxDate = props.maxDate,
        maxDateProp = _props$maxDate === void 0 ? defaultMaxDate : _props$maxDate,
        other = _objectWithoutProperties(props, ["calendars", "value", "onChange", "mask", "startText", "endText", "inputFormat", "minDate", "maxDate"]);

    var utils = useUtils();
    var minDate = useParsedDate(minDateProp);
    var maxDate = useParsedDate(maxDateProp);

    var _React$useState = React.useState('start'),
        currentlySelectingRangeEnd = _React$useState[0],
        setCurrentlySelectingRangeEnd = _React$useState[1];

    var pickerStateProps = _extends({}, other, {
      value: value,
      onChange: onChange,
      inputFormat: passedInputFormat || utils.formats.keyboardDate
    });

    var restProps = _extends({}, other, {
      minDate: minDate,
      maxDate: maxDate
    });

    var _usePickerState = usePickerState(pickerStateProps, rangePickerValueManager),
        pickerProps = _usePickerState.pickerProps,
        inputProps = _usePickerState.inputProps,
        wrapperProps = _usePickerState.wrapperProps;

    var validationError = useDateRangeValidation(value, restProps);

    var DateInputProps = _extends({}, inputProps, restProps, {
      currentlySelectingRangeEnd: currentlySelectingRangeEnd,
      setCurrentlySelectingRangeEnd: setCurrentlySelectingRangeEnd,
      startText: startText,
      endText: endText,
      mask: mask,
      validationError: validationError,
      ref: ref
    });

    return /*#__PURE__*/_jsx(WrapperComponent, _extends({
      wrapperProps: wrapperProps,
      DateInputProps: DateInputProps
    }, restProps, {
      children: /*#__PURE__*/_jsx(DateRangePickerView, _extends({
        open: wrapperProps.open,
        DateInputProps: DateInputProps,
        calendars: calendars,
        currentlySelectingRangeEnd: currentlySelectingRangeEnd,
        setCurrentlySelectingRangeEnd: setCurrentlySelectingRangeEnd,
        startText: startText,
        endText: endText
      }, pickerProps, restProps))
    }));
  }); // @ts-expect-error Impossible to save component generics when wrapping with HOC

  return RangePickerWithStateAndWrapper;
}
